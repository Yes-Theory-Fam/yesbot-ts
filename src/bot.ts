import {
  Client,
  GuildMember,
  Message,
  MessageReaction,
  PartialGuildMember,
  PartialUser,
  User,
  VoiceState,
} from "discord.js";
import {
  GuildMemberUpdate,
  MemberJoin,
  MemberLeave,
  MessageManager,
  ReactionAdd,
  ReactionRemove,
  Ready,
  VoiceStateUpdate,
} from "./events";
import { createYesBotLogger } from "./log";

const logger = createYesBotLogger("main", "Bot");

export class Bot {
  private static instance: Bot;
  private readonly client: Client;

  constructor() {
    this.client = new Client({ partials: ["REACTION", "MESSAGE"] });
    this.client.login(process.env.BOT_TOKEN).then(r => this.createEventListeners());
    logger.debug("Logging in to Discord Gateway");
  }

  public static getInstance(): Bot {
    if (!Bot.instance) {
      Bot.instance = new Bot();
    }

    return Bot.instance;
  }

  public getClient(): Client {
    return this.client;
  }

  private createEventListeners() {
    this.client.on(
      "guildMemberAdd",
      (member: GuildMember | PartialGuildMember) => new MemberJoin(member),
    );
    this.client.on(
      "guildMemberRemove",
      (member: GuildMember | PartialGuildMember) => new MemberLeave(member),
    );
    this.client.on(
      "guildMemberUpdate",
      (
        oldMember: GuildMember | PartialGuildMember,
        newMember: GuildMember | PartialGuildMember,
      ) => new GuildMemberUpdate(oldMember, newMember),
    );
    this.client.on("message", (msg: Message) => new MessageManager(msg));
    this.client.on(
      "messageReactionAdd",
      (messageReaction: MessageReaction, user: User | PartialUser) =>
        new ReactionAdd(messageReaction, user),
    );
    this.client.on(
      "messageReactionRemove",
      (messageReaction: MessageReaction, user: User | PartialUser) =>
        new ReactionRemove(messageReaction, user),
    );
    this.client.on("ready", () => new Ready(this.client));
    this.client.on(
      "voiceStateUpdate",
      (oldMember: VoiceState, newMember: VoiceState) =>
        new VoiceStateUpdate(oldMember, newMember),
    );
  }
}
