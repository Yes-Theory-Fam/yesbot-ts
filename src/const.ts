export const GUILD_ID = 449984633342525462;
export const CHAT_ID = 449984633908625409;
export const BERT_ID = 283678513654136842;
export const JAMIE_ID = 219502124709445633;
export const BOT_DEV_ID = 636147342512357398;
export const BOT_PROD_ID = 614101602046836776;
export const BOT_PROD_TOKEN = "NjE0MTAxNjAyMDQ2ODM2Nzc2.XiIw2Q.aaVUHwW8y31cGkEohn1dnukMAaQ";
export const BOT_DEV_TOKEN = "NjM2MTQ3MzQyNTEyMzU3Mzk4.XiJVcA.hB0AbneEX00HD0DSdDTIybTylPs";
export const MODERATOR_ROLE_NAME = "Support";
export const ENGINEER_ROLE_NAME = "Server Engineer";
export const COORDINATOR_ROLE_NAME = "Server Coordinator";