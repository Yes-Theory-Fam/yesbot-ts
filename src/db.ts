// Imported for side-effects. Used by typeorm
import "reflect-metadata";
import { createConnection } from "typeorm";
import { createYesBotLogger } from "./log";

const logger = createYesBotLogger("db", "init");

createConnection().then(async _ => {
    logger.info("Database connection established.");
}).catch(err => {
    logger.error("Database connection failed: ", err);
})
