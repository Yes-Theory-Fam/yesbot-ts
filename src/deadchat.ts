class Deadchat {

    classProperty: string;

    constructor(content: string) {
        this.classProperty = content
        
    }
}





export default Deadchat;