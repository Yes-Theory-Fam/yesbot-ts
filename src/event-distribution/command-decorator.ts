import { EventHandlerOptions } from "./events/events";
import { BaseOptions, DiscordEvent, EventLocation } from "./types/base";
import { HandlerClass } from "./types/handler";
import distribution from "./index";
import { createYesBotLogger } from "../log";

const logger = createYesBotLogger("event-distribution", "Command");
const explanation =
  "Commands that require roles or channel names won't work in DMs since the roles cannot be read from DM events.";

export const Command = <T extends EventHandlerOptions>(options: T) => {
  if (
    options.event !== DiscordEvent.GUILD_MEMBER_UPDATE &&
    options.event !== DiscordEvent.MEMBER_LEAVE &&
    options.event !== DiscordEvent.VOICE_STATE_UPDATE &&
    options.event !== DiscordEvent.TIMER &&
    options.event !== DiscordEvent.READY
  ) {
    setDefaultOnBaseOptions(options);
  }

  return <U extends HandlerClass<T["event"]>>(target: U) => {
    const commandClassName = target.name;
    logger.debug(`Loading new command: ${target.name}`);

    if (
      options.event !== DiscordEvent.GUILD_MEMBER_UPDATE &&
      options.event !== DiscordEvent.MEMBER_LEAVE &&
      options.event !== DiscordEvent.VOICE_STATE_UPDATE &&
      options.event !== DiscordEvent.TIMER &&
      options.event !== DiscordEvent.READY
    ) {
      checkBaseOptions(options, commandClassName);
    }

    try {
      distribution.addWithOptions(options, target);
    } catch (e) {
      logger.error(
        `Failed to load command ${commandClassName}! Error was: `,
        e instanceof Error ? { error: e.message } : e
      );
    }
    return target;
  };
};

const baseOptionsRequireServer = (options: BaseOptions) => {
  const channels = options.channelNames ?? [];
  const roles = options.requiredRoles ?? [];
  return channels.length > 0 || roles.length > 0;
};

const setDefaultOnBaseOptions = (options: BaseOptions) => {
  options.location ??= baseOptionsRequireServer(options)
    ? EventLocation.SERVER
    : EventLocation.ANYWHERE;

  options.requiredRoles ??= [];
};

const checkBaseOptions = (options: BaseOptions, commandClassName: string) => {
  const requiresServer = baseOptionsRequireServer(options);
  if (requiresServer && options.location === EventLocation.DIRECT_MESSAGE) {
    logger.error(
      `Failed to load command ${commandClassName}! The command requires roles or channel names but was set to location ${EventLocation.DIRECT_MESSAGE}. ${explanation}`
    );
    return;
  }

  if (requiresServer && options.location === EventLocation.ANYWHERE) {
    logger.warn(
      `Adding command ${commandClassName} with required roles or limited channels which was set to location ${EventLocation.ANYWHERE}. ${explanation}`
    );
  }
};
