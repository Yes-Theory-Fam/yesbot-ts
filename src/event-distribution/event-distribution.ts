import {
  addEventHandler,
  EventHandlerOptions,
  extractEventInfo,
  HandlerFunction,
} from "./events/events";
import glob from "glob";
import path from "path";
import { createYesBotLogger } from "../log";
import {
  HIOC,
  isHIOCArray,
  StringIndexedHIOCTree,
  StringIndexedHIOCTreeNode,
} from "./types/hioc";
import { HandlerClass } from "./types/handler";
import { DiscordEvent, EventLocation } from "./types/base";

const logger = createYesBotLogger("event-distribution", "event-distribution");

type EventDistributionHandlers = {
  [key in DiscordEvent]: StringIndexedHIOCTree<key>;
};

export class EventDistribution {
  handlers: EventDistributionHandlers = {
    [DiscordEvent.MEMBER_LEAVE]: {},
    [DiscordEvent.MESSAGE]: {},
    [DiscordEvent.REACTION_ADD]: {},
    [DiscordEvent.REACTION_REMOVE]: {},
    [DiscordEvent.GUILD_MEMBER_UPDATE]: {},
    [DiscordEvent.READY]: {},
    [DiscordEvent.VOICE_STATE_UPDATE]: {},
  };

  handleEvent<T extends DiscordEvent>(
    event: T,
    ...args: Parameters<HandlerFunction<T>>
  ) {
    const infos = extractEventInfo(event, ...args);

    for (const info of infos) {
      const { handlerKeys, isDirectMessage, member } = info;

      const roleNames = member?.roles.cache.map((r) => r.name) ?? [];
      const eventHandlers = this.getHandlers(this.handlers[event], handlerKeys);
      const filteredHandlers = this.filterHandlers(
        eventHandlers,
        isDirectMessage,
        roleNames
      );

      for (const { ioc } of filteredHandlers) {
        let instance = ioc;
        if (typeof instance === "function") instance = new instance();

        instance.handle(...args);
      }
    }
  }

  addWithOptions<T extends EventHandlerOptions>(
    options: T,
    HandlerClass: HandlerClass<T["event"]>
  ) {
    const ioc = options.stateful ? new HandlerClass() : HandlerClass;
    const tree = this.handlers[options.event];
    addEventHandler(options, ioc, tree);
  }

  async initialize(): Promise<void> {
    return new Promise((res, rej) => {
      const extension = process.env.NODE_ENV === "production" ? ".js" : ".ts";
      const directory = process.env.NODE_ENV === "production" ? "build" : "src";

      glob(`${directory}/programs/*${extension}`, async (e, matches) => {
        if (e) {
          logger.error("Error loading commands: ", e);
          return;
        }

        const loaders = matches
          .filter((p) => !p.endsWith(`.spec${extension}`))
          .map((p) => {
            const split = p.split(".");
            split.unshift();
            const modulePath = path.join(process.cwd(), split.join("."));

            return import(modulePath);
          });

        try {
          await Promise.all(loaders);
        } catch (e) {
          logger.error("Error loading commands: ", e);
          rej(e);
        }
        logger.debug("Loading complete!");
        res();
      });
    });
  }

  private filterHandlers<T extends DiscordEvent>(
    handlers: HIOC<T>[],
    isDirectMessage: boolean,
    roleNames: string[]
  ): HIOC<T>[] {
    const locationFilteredHandlers = handlers.filter((eh) => {
      if (
        eh.options.event === DiscordEvent.GUILD_MEMBER_UPDATE ||
        eh.options.event === DiscordEvent.MEMBER_LEAVE ||
        eh.options.event === DiscordEvent.VOICE_STATE_UPDATE ||
        eh.options.event === DiscordEvent.READY
      )
        return true;

      const { location } = eh.options;
      switch (location) {
        case EventLocation.ANYWHERE:
          return true;
        case EventLocation.DIRECT_MESSAGE:
          return isDirectMessage;
        case EventLocation.SERVER:
          return !isDirectMessage;
      }
    });

    return locationFilteredHandlers.filter((eh) => {
      if (
        eh.options.event === DiscordEvent.GUILD_MEMBER_UPDATE ||
        eh.options.event === DiscordEvent.MEMBER_LEAVE ||
        eh.options.event === DiscordEvent.VOICE_STATE_UPDATE ||
        eh.options.event === DiscordEvent.READY
      )
        return true;

      const { requiredRoles } = eh.options;
      return requiredRoles.every((role) => roleNames.includes(role));
    });
  }

  private getHandlers<T extends DiscordEvent>(
    handlerTree: StringIndexedHIOCTreeNode<T>,
    [currentKey, ...restKeys]: string[]
  ): HIOC<T>[] {
    if (!handlerTree) return [];

    if (isHIOCArray(handlerTree)) return handlerTree;

    const handlers: HIOC<T>[] = [];
    handlers.push(...this.getHandlers(handlerTree[""], restKeys));

    if (!currentKey) {
      return handlers;
    }

    handlers.push(...this.getHandlers(handlerTree[currentKey], restKeys));

    return handlers;
  }
}
