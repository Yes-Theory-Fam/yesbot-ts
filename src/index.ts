import { createYesBotLogger } from "./log";
import {
  Client,
  GuildMember,
  Message,
  MessageReaction,
  PartialGuildMember,
  PartialUser,
  User,
  VoiceState,
} from "discord.js";
import {
  guildMemberUpdate,
  memberLeave,
  messageManager,
  reactionAdd,
  reactionRemove,
  ready,
  voiceStateUpdate,
} from "./events";
import distribution, { DiscordEvent } from "./event-distribution";

const logger = createYesBotLogger("main", "index");
logger.info("Starting YesBot");

const bot = new Client({ partials: ["REACTION", "MESSAGE"] });
logger.info("Initializing event-distribution");
distribution.initialize().then(() => {
  logger.debug("Logging in to Discord Gateway");
  return bot.login(process.env.BOT_TOKEN);
});

//! ================= EVENT HANDLERS ====================
bot.on(
  "guildMemberRemove",
  async (member: GuildMember | PartialGuildMember) => {
    distribution.handleEvent(DiscordEvent.MEMBER_LEAVE, member);
    await memberLeave(member);
  }
);
bot.on("guildMemberAdd", async (member: GuildMember | PartialGuildMember) => {
  distribution.handleEvent(DiscordEvent.MEMBER_JOIN, member);
});
bot.on(
  "guildMemberUpdate",
  async (
    oldMember: GuildMember | PartialGuildMember,
    newMember: GuildMember | PartialGuildMember
  ) => {
    await guildMemberUpdate(oldMember, newMember);
    distribution.handleEvent(
      DiscordEvent.GUILD_MEMBER_UPDATE,
      oldMember,
      newMember
    );
  }
);
bot.on("message", async (msg: Message) => {
  await messageManager(msg);
  distribution.handleEvent(DiscordEvent.MESSAGE, msg);
});
bot.on(
  "messageReactionAdd",
  async (messageReaction: MessageReaction, user: User | PartialUser) => {
    await reactionAdd(messageReaction, user);
    distribution.handleEvent(DiscordEvent.REACTION_ADD, messageReaction, user);
  }
);
bot.on(
  "messageReactionRemove",
  async (messageReaction: MessageReaction, user: User | PartialUser) => {
    await reactionRemove(messageReaction, user);
    distribution.handleEvent(
      DiscordEvent.REACTION_REMOVE,
      messageReaction,
      user
    );
  }
);
bot.on("ready", async () => {
  distribution.handleEvent(DiscordEvent.READY, bot);
  await ready(bot);
});
bot.on(
  "voiceStateUpdate",
  async (oldState: VoiceState, newState: VoiceState) => {
    distribution.handleEvent(
      DiscordEvent.VOICE_STATE_UPDATE,
      oldState,
      newState
    );
    await voiceStateUpdate(oldState, newState);
  }
);
//! ================= /EVENT HANDLERS ===================

export default bot;
module.exports = bot; // Required for require() when lazy loading.
