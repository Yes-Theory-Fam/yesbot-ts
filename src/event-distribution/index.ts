import { EventDistribution } from "./event-distribution.js";

const distribution = new EventDistribution();
export default distribution;

export * from "./types/handler.js";
export * from "./types/base.js";
export * from "./command-decorator.js";

export { DiscordEvent } from "./types/base.js";
