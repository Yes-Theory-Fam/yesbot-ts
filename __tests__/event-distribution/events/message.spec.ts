import { ChannelType } from "discord.js";
import {
  addMessageHandler,
  extractMessageInfo,
  MessageEventHandlerOptions,
} from "../../../src/event-distribution/events/message";
import MockDiscord from "../../mocks";
import {
  InstanceOrConstructor,
  StringIndexedHIOCTree,
} from "../../../src/event-distribution/types/hioc";
import { CommandHandler, DiscordEvent } from "../../../src/event-distribution";

describe("Messages", () => {
  let mockDiscord: MockDiscord;
  beforeEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should extract messages from public channel", () => {
    const message = mockDiscord.getMessage();
    const result = extractMessageInfo(message);
    expect(result).toMatchSnapshot();
  });

  it("should extract messages from direct message", () => {
    const message = mockDiscord.getMessage();
    message.channel.type = ChannelType.DM;
    const result = extractMessageInfo(message);
    expect(result).toMatchSnapshot();
  });

  it("should add data to the tree", () => {
    const tree: StringIndexedHIOCTree<DiscordEvent.MESSAGE> = { channel: {} };
    const options: MessageEventHandlerOptions = {
      description: "test",
      event: DiscordEvent.MESSAGE,
      trigger: "!test",
      channelNames: ["bot-output"],
    };
    const ioc = {} as InstanceOrConstructor<
      CommandHandler<DiscordEvent.MESSAGE>
    >;

    addMessageHandler(options, ioc, tree);
    expect(tree).toMatchSnapshot();

    const secondOptions = { ...options, allowedRoles: ["Support"] };
    addMessageHandler(secondOptions, ioc, tree);
    expect(tree).toMatchSnapshot();

    const thirdOptions = {
      ...options,
      trigger: "!otherTrigger",
      allowedRoles: ["Developer"],
      channelNames: ["bot-development", "bot-test-channel"],
    };
    addMessageHandler(thirdOptions, ioc, tree);
    expect(tree).toMatchSnapshot();
  });
});
