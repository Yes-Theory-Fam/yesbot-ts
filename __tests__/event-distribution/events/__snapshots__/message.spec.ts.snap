// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages should add data to the tree 1`] = `
{
  "bot-output": {
    "!test": {
      "": [
        {
          "ioc": {},
          "options": {
            "channelNames": [
              "bot-output",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!test",
          },
        },
      ],
    },
  },
  "channel": {},
}
`;

exports[`Messages should add data to the tree 2`] = `
{
  "bot-output": {
    "!test": {
      "": [
        {
          "ioc": {},
          "options": {
            "channelNames": [
              "bot-output",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!test",
          },
        },
        {
          "ioc": {},
          "options": {
            "allowedRoles": [
              "Support",
            ],
            "channelNames": [
              "bot-output",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!test",
          },
        },
      ],
    },
  },
  "channel": {},
}
`;

exports[`Messages should add data to the tree 3`] = `
{
  "bot-development": {
    "!othertrigger": {
      "": [
        {
          "ioc": {},
          "options": {
            "allowedRoles": [
              "Developer",
            ],
            "channelNames": [
              "bot-development",
              "bot-test-channel",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!otherTrigger",
          },
        },
      ],
    },
  },
  "bot-output": {
    "!test": {
      "": [
        {
          "ioc": {},
          "options": {
            "channelNames": [
              "bot-output",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!test",
          },
        },
        {
          "ioc": {},
          "options": {
            "allowedRoles": [
              "Support",
            ],
            "channelNames": [
              "bot-output",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!test",
          },
        },
      ],
    },
  },
  "bot-test-channel": {
    "!othertrigger": {
      "": [
        {
          "ioc": {},
          "options": {
            "allowedRoles": [
              "Developer",
            ],
            "channelNames": [
              "bot-development",
              "bot-test-channel",
            ],
            "description": "test",
            "event": "MESSAGE",
            "trigger": "!otherTrigger",
          },
        },
      ],
    },
  },
  "channel": {},
}
`;

exports[`Messages should extract messages from direct message 1`] = `
[
  {
    "content": "this is the message content",
    "handlerKeys": [
      "9007199254740991",
      "this",
      "is",
    ],
    "isDirectMessage": true,
    "member": {
      "avatar": null,
      "avatarURL": null,
      "communicationDisabledUntilTimestamp": null,
      "displayAvatarURL": "https://cdn.discordapp.com/avatars/9007199254740991/user%20avatar%20url.webp",
      "displayName": "nick",
      "flags": 0,
      "guildId": "9007199254740991",
      "joinedTimestamp": NaN,
      "nickname": "nick",
      "pending": false,
      "premiumSinceTimestamp": null,
      "roles": {
        "cache": [],
      },
      "userId": "9007199254740991",
    },
  },
]
`;

exports[`Messages should extract messages from public channel 1`] = `
[
  {
    "content": "this is the message content",
    "handlerKeys": [
      "9007199254740991",
      "this",
      "is",
    ],
    "isDirectMessage": false,
    "member": {
      "avatar": null,
      "avatarURL": null,
      "communicationDisabledUntilTimestamp": null,
      "displayAvatarURL": "https://cdn.discordapp.com/avatars/9007199254740991/user%20avatar%20url.webp",
      "displayName": "nick",
      "flags": 0,
      "guildId": "9007199254740991",
      "joinedTimestamp": NaN,
      "nickname": "nick",
      "pending": false,
      "premiumSinceTimestamp": null,
      "roles": {
        "cache": [],
      },
      "userId": "9007199254740991",
    },
  },
]
`;
